packages/job-manager/src/batch/batch.ts

```diff
@@ -26,133 +25,139 @@ export class Batch {
+                    serviceJobId = `${jobId}_${crypto.randomBytes(5).toString('hex')}`;
                 }
-            });
+
+                const jobAddParameter: BatchServiceModels.JobAddParameter = {
+                    id: serviceJobId,
+                    poolInfo: {
+                        poolId: this.config.poolId,
+                    },
+                    onAllTasksComplete: 'terminatejob',
+                };
+
+                await client.job.add(jobAddParameter);
+
+                this.logger.logInfo(`New job ${serviceJobId} created.`);
+            } else {
+                throw new VError(error as Error, `An error occurred while retrieving state of ${jobId} job.`);
+            }
+        }
 
         return serviceJobId;
     }
 
     public async createTasks(jobId: string, messages: Message[]): Promise<JobTask[]> {
-        const taskAddParameters: BatchServiceModels.TaskAddParameter[] = [];
-        const maxTaskDurationInMinutes = await this.getMaxTaskDurationInMinutes();
+        // Azure Batch supports the maximum 100 tasks to add in collection
+        const tasks: JobTask[] = [];
 
-        messages.forEach(message => {
-            const jobTask = new JobTask(message.messageId);
-            this.jobTasks.set(jobTask.id, jobTask);
-            const taskAddParameter = this.getTaskAddParameter(jobTask.id, message.messageText, maxTaskDurationInMinutes);
-            taskAddParameters.push(taskAddParameter);
-        });
+        let position = 0;
+        let count = messages.length > 100 ? 100 : messages.length;
+        do {
+            const taskCollection = await this.addTaskCollection(jobId, messages.slice(position, position + count));
+            tasks.push(...taskCollection);
 
-        if (taskAddParameters.length > 0) {
-            const client = await this.batchClientProvider();
-            const taskAddCollectionResult = await client.task.addCollection(jobId, taskAddParameters);
-            taskAddCollectionResult.value.forEach(taskAddResult => {
-                if (/success/i.test(taskAddResult.status)) {
-                    this.jobTasks.get(taskAddResult.taskId).state = JobTaskState.queued;
-                    this.logger.logInfo(`New task ${taskAddResult.taskId} added to the job ${jobId}.`);
-                } else {
-                    this.jobTasks.get(taskAddResult.taskId).state = JobTaskState.failed;
-                    this.jobTasks.get(taskAddResult.taskId).error = taskAddResult.error.message.value;
-                    this.logger.logError(`An error occurred while adding new task ${JSON.stringify(taskAddResult)} to the job ${jobId}.`);
-                }
-            });
-        } else {
-            this.logger.logInfo(`No new tasks added to the job ${jobId}.`);
-        }
+            position = position + count;
+            count = messages.length - position > 100 ? 100 : messages.length - position;
+        } while (position < messages.length);
```

Comment:
how about creating a helper function in common:

export function getChunks<T>(allItems: T[], chunkSize: number): T[][] {
        const chunks: T[][] = [];
        const length = allItems.length;

        for (let chunkPos = 0; chunkPos < length; chunkPos += chunkSize) {
            chunks.push(allItems.slice(chunkPos, chunkPos + chunkSize));
        }

        return chunks;
    }

Reply:
& then iterate on each chunk, which may be more readable?

---
In reply to: [308975673](https://github.com/microsoft/accessibility-insights-service/pull/222#discussion_r308975673) [](ancestors = 308975673)

Reply:
moved to common